pco_c/src/lib.rs

```diff
@@ -0,0 +1,111 @@
+use libc::{c_uint, c_void};
+use pco::data_types::NumberLike;
+
+#[repr(C)]
+pub struct PcoVec {
+  ptr: *const c_void,
+  len: c_uint,
+  raw_box: *const c_void,
+}
+
+#[repr(C)]
+pub enum PcoType {
+  U8,
+  I32,
+  I64,
+  U32,
+  U64,
+  F32,
+  F64,
+}
+
+impl PcoVec {
+  pub fn from_vec<T: Sized>(v: Vec<T>) -> Self {
+    let ptr = v.as_ptr();
+    let len = v.len() as c_uint;
+    let raw_box = Box::into_raw(Box::new(v));
+    PcoVec {
+      ptr: ptr as *const c_void,
+      len,
+      raw_box: raw_box as *const c_void,
+    }
+  }
+
+  pub fn free<T: Sized>(self) {
+    unsafe { drop(Box::from_raw(self.raw_box as *mut Vec<T>)) }
+  }
+
+  pub fn empty() -> Self {
+    PcoVec {
+      ptr: std::ptr::null(),
+      len: 0,
+      raw_box: std::ptr::null(),
+    }
+  }
+}
+
+fn _auto_compress<T: NumberLike>(nums: *const c_void, len: c_uint, level: c_uint) -> PcoVec {
+  let slice = unsafe { std::slice::from_raw_parts(nums as *const T, len as usize) };
+  PcoVec::from_vec(pco::standalone::auto_compress(
+    slice,
+    level as usize,
+  ))
+}
+
+fn _auto_decompress<T: NumberLike>(compressed: *const u8, len: c_uint) -> PcoVec {
+  let slice = unsafe { std::slice::from_raw_parts(compressed, len as usize) };
+  let decompressed = pco::standalone::auto_decompress::<T>(slice);
+  match decompressed {
+    Ok(v) => PcoVec::from_vec(v),
+    Err(_) => PcoVec::empty(),
```

Comment:
I think we should surface an error in this case, instead of falsely returning an empty vector. I'm not super familiar with C, but perhaps it would make sense to add an error code to `PcoVec`? Alternatively, even panicking by calling `.unwrap()` on the Rust side would be good enough for now.

Reply:
The typical thing to do would be to change the return values of all the FFI functions to return ints, have zero be success and other values represent failures defined in an enum. The structs would then instead be passed via pointers to Rust.

Reply:
That sounds good to me.